File: test/cpu/move.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;
BOOST_FIXTURE_TEST_SUITE(MOVE, Prepare)
struct F {
    F() {
        // MOVE.B %D2, %D3
        TEST::SET_W(0, 0010000 | 3 << 9 | 2);
        TEST::SET_W(2, TEST_BREAK);

        // MOVE.B (%A2), %D3
        TEST::SET_W(4, 0010020 | 3 << 9 | 2);
        TEST::SET_W(6, TEST_BREAK);

        // MOVE.B (%A2+), %D3
        TEST::SET_W(8, 0010030 | 3 << 9 | 2);
        TEST::SET_W(10, TEST_BREAK);

        // MOVE.B (%A7+), %D3
        TEST::SET_W(12, 0010030 | 3 << 9 | 7);
        TEST::SET_W(14, TEST_BREAK);

        // MOVE.B (-%A2), %D3
        TEST::SET_W(16, 0010040 | 3 << 9 | 2);
        TEST::SET_W(18, TEST_BREAK);

        // MOVE.B (-%A7), %D3
        TEST::SET_W(20, 0010040 | 3 << 9 | 7);
        TEST::SET_W(22, TEST_BREAK);

        // MOVE.B #0x34, %D3
        TEST::SET_W(24, 0010074 | 3 << 9);
        TEST::SET_W(26, 0x34);
        TEST::SET_W(28, TEST_BREAK);

        // MOVE.W %D2, %D3
        TEST::SET_W(30, 0030000 | 3 << 9 | 2);
        TEST::SET_W(32, TEST_BREAK);

        // MOVE.W %A2, %D3
        TEST::SET_W(34, 0030010 | 3 << 9 | 2);
        TEST::SET_W(36, TEST_BREAK);

        // MOVE.W (%A2), %D3
        TEST::SET_W(38, 0030020 | 3 << 9 | 2);
        TEST::SET_W(40, TEST_BREAK);

        // MOVE.W (%A2+), %D3
        TEST::SET_W(42, 0030030 | 3 << 9 | 2);
        TEST::SET_W(44, TEST_BREAK);

        // MOVE.W (-%A2), %D3
        TEST::SET_W(46, 0030040 | 3 << 9 | 2);
        TEST::SET_W(48, TEST_BREAK);

        // MOVE.W #0x3456, %D3
        TEST::SET_W(50, 0030074 | 3 << 9);
        TEST::SET_W(52, 0x3456);
        TEST::SET_W(54, TEST_BREAK);

        // MOVE.W %CCR, %D3
        TEST::SET_W(56, 0041300 | 3);
        TEST::SET_W(58, TEST_BREAK);

        // MOVE.W %SR, %D3
        TEST::SET_W(60, 0040300 | 3);
        TEST::SET_W(62, TEST_BREAK);

        // MOVE.L %D2, %D3
        TEST::SET_W(64, 0020000 | 3 << 9 | 2);
        TEST::SET_W(66, TEST_BREAK);

        // MOVE.L %A2, %D3
        TEST::SET_W(68, 0020010 | 3 << 9 | 2);
        TEST::SET_W(70, TEST_BREAK);

        // MOVE.L (%A2), %D3
        TEST::SET_W(72, 0020020 | 3 << 9 | 2);
        TEST::SET_W(74, TEST_BREAK);

        // MOVE.L (%A2+), %D3
        TEST::SET_W(76, 0020030 | 3 << 9 | 2);
        TEST::SET_W(78, TEST_BREAK);

        // MOVE.L (-%A2), %D3
        TEST::SET_W(80, 0020040 | 3 << 9 | 2);
        TEST::SET_W(82, TEST_BREAK);

        // MOVE.L #0x3456789A, %D3
        TEST::SET_W(84, 0020074 | 3 << 9);
        TEST::SET_L(86, 0x3456789A);
        TEST::SET_W(90, TEST_BREAK);

        // MOVE %USP, %A3
        TEST::SET_W(92, 0047150 | 3);
        TEST::SET_W(94, TEST_BREAK);

        // MOVE.B %D3, (%A2)
        TEST::SET_W(96, 0010200 | 2 << 9 | 3);
        TEST::SET_W(98, TEST_BREAK);

        // MOVE.B (0x100), (0x200)
        TEST::SET_W(100, 0011771);
        TEST::SET_L(102, 0x100);
        TEST::SET_L(106, 0x200);
        TEST::SET_W(110, TEST_BREAK);

        // MOVE.W %D3, (%A2)
        TEST::SET_W(112, 0030200 | 2 << 9 | 3);
        TEST::SET_W(114, TEST_BREAK);

        // MOVE.W (0x100), (0x200)
        TEST::SET_W(116, 0031771);
        TEST::SET_L(118, 0x100);
        TEST::SET_L(122, 0x200);
        TEST::SET_W(126, TEST_BREAK);

        // MOVE.W %D3, %CCR
        TEST::SET_W(128, 0042300 | 3);
        TEST::SET_W(130, TEST_BREAK);

        // MOVE.W %D3, %SR
        TEST::SET_W(132, 0043300 | 3);
        TEST::SET_W(134, TEST_BREAK);

        // MOVE.L %D3, (%A2)
        TEST::SET_W(136, 0020200 | 2 << 9 | 3);
        TEST::SET_W(138, TEST_BREAK);

        // MOVE.L (0x100), (0x200)
        TEST::SET_W(140, 0021771);
        TEST::SET_L(142, 0x100);
        TEST::SET_L(146, 0x200);
        TEST::SET_W(150, TEST_BREAK);

        // MOVE %A3, %USP
        TEST::SET_W(152, 0047140 | 3);
        TEST::SET_W(154, TEST_BREAK);

        jit_compile(0, 156);
    }
};
BOOST_AUTO_TEST_SUITE(R, *boost::unit_test::fixture<F>())
BOOST_AUTO_TEST_SUITE(Load)
BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(dr) {
    cpu.D[3] = 0x12345678;
    cpu.D[2] = 0x9ABCDE02;
    cpu.C = cpu.V = cpu.X = true;
    run_test(0);
    BOOST_TEST(cpu.D[3] == 0x12345602);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x80;
    run_test(0);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(0);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_CASE(mem) {
    RAM[0x100] = 0x02;
    cpu.A[2] = 0x100;
    run_test(4);
    BOOST_TEST(cpu.D[3] == 0x02);
}

BOOST_AUTO_TEST_CASE(inc) {
    RAM[0x100] = 0x02;
    cpu.A[2] = 0x100;
    run_test(8);
    BOOST_TEST(cpu.D[3] == 0x02);
    BOOST_TEST(cpu.A[2] == 0x101);
}

BOOST_AUTO_TEST_CASE(pop) {
    RAM[0x100] = 0x02;
    cpu.A[7] = 0x100;
    run_test(12);
    BOOST_TEST(cpu.D[3] == 0x02);
    BOOST_TEST(cpu.A[7] == 0x102);
}

BOOST_AUTO_TEST_CASE(dec) {
    RAM[0x100] = 0x02;
    cpu.D[3] = 0x12345678;
    cpu.A[2] = 0x101;
    run_test(16);
    BOOST_TEST(cpu.D[3] == 0x12345602);
    BOOST_TEST(cpu.A[2] == 0x100);
}

BOOST_AUTO_TEST_CASE(push) {
    RAM[0x100] = 0x02;
    cpu.A[7] = 0x102;
    run_test(20);
    BOOST_TEST(cpu.D[3] == 0x02);
    BOOST_TEST(cpu.A[7] == 0x100);
}

BOOST_AUTO_TEST_CASE(imm) {
    cpu.D[3] = 0x12345678;
    run_test(24);
    BOOST_TEST(cpu.D[3] == 0x12345634);
}

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(dr) {
    cpu.D[3] = 0x12345678;
    cpu.D[2] = 0x9ABC1234;
    cpu.C = cpu.V = cpu.X = true;
    run_test(30);
    BOOST_TEST(cpu.D[3] == 0x12341234);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x8000;
    run_test(30);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(30);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_CASE(ar) {
    cpu.D[3] = 0x12345678;
    cpu.A[2] = 0x9ABCDE02;
    run_test(34);
    BOOST_TEST(cpu.D[3] == 0x1234DE02);
}

BOOST_AUTO_TEST_CASE(mem) {
    TEST::SET_W(0x100, 0xabcd);
    cpu.A[2] = 0x100;
    run_test(38);
    BOOST_TEST(cpu.D[3] == 0xabcd);
}

BOOST_AUTO_TEST_CASE(inc) {
    TEST::SET_W(0x100, 0xabcd);
    cpu.A[2] = 0x100;
    run_test(42);
    BOOST_TEST(cpu.D[3] == 0xabcd);
    BOOST_TEST(cpu.A[2] == 0x102);
}

BOOST_AUTO_TEST_CASE(dec) {
    TEST::SET_W(0x100, 0xabcd);
    cpu.A[2] = 0x102;
    run_test(46);
    BOOST_TEST(cpu.D[3] == 0xabcd);
    BOOST_TEST(cpu.A[2] == 0x100);
}

BOOST_AUTO_TEST_CASE(imm) {
    cpu.D[3] = 0x12345678;
    run_test(50);
    BOOST_TEST(cpu.D[3] == 0x12343456);
}

BOOST_AUTO_TEST_CASE(CCR) {
    cpu.X = cpu.N = cpu.Z = cpu.V = cpu.C = true;
    run_test(56);
    BOOST_TEST(cpu.D[3] == 0x1F);
}

BOOST_AUTO_TEST_SUITE(SR)
BOOST_AUTO_TEST_CASE(NG) {
    cpu.S = false;
    run_test(60);
    BOOST_TEST(cpu.ex_n == 8);
}

BOOST_AUTO_TEST_CASE(OK) {
    cpu.S = true;
    cpu.T = 0;
    cpu.M = false;
    cpu.X = cpu.N = cpu.Z = cpu.V = cpu.C = true;
    run_test(60);
    BOOST_TEST(cpu.D[3] == 0x201F);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(dr) {
    cpu.D[2] = 0x1ABC1234;
    cpu.C = cpu.V = cpu.X = true;
    run_test(64);
    BOOST_TEST(cpu.D[3] == 0x1ABC1234);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.V);
    BOOST_TEST(cpu.X);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.Z);
}

BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = 0x80000000;
    run_test(64);
    BOOST_TEST(cpu.N);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = 0;
    run_test(64);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_CASE(ar) {
    cpu.A[2] = 0x9ABCDE02;
    run_test(68);
    BOOST_TEST(cpu.D[3] == 0x9ABCDE02);
}

BOOST_AUTO_TEST_CASE(mem) {
    TEST::SET_L(0x100, 0xabcd0123);
    cpu.A[2] = 0x100;
    run_test(72);
    BOOST_TEST(cpu.D[3] == 0xabcd0123);
}

BOOST_AUTO_TEST_CASE(inc) {
    TEST::SET_L(0x100, 0xabcd0123);
    cpu.A[2] = 0x100;
    run_test(76);
    BOOST_TEST(cpu.D[3] == 0xabcd0123);
    BOOST_TEST(cpu.A[2] == 0x104);
}

BOOST_AUTO_TEST_CASE(dec) {
    TEST::SET_L(0x100, 0xabcd0123);
    cpu.A[2] = 0x104;
    run_test(80);
    BOOST_TEST(cpu.D[3] == 0xabcd0123);
    BOOST_TEST(cpu.A[2] == 0x100);
}

BOOST_AUTO_TEST_CASE(imm) {
    cpu.D[3] = 0x12345678;
    run_test(84);
    BOOST_TEST(cpu.D[3] == 0x3456789A);
}

BOOST_AUTO_TEST_SUITE(USP)

BOOST_AUTO_TEST_CASE(NG) {
    cpu.S = false;
    run_test(92);
    BOOST_TEST(cpu.ex_n == 8);
}

BOOST_AUTO_TEST_CASE(OK) {
    cpu.S = true;
    cpu.USP = 0x300;
    run_test(92);
    BOOST_TEST(cpu.A[3] == 0x300);
}

BOOST_AUTO_TEST_CASE(Traced) {
    cpu.S = true;
    cpu.T = 1;
    run_test(92);
    BOOST_TEST(cpu.ex_n == 9);
}
BOOST_AUTO_TEST_SUITE_END()

BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Store)
BOOST_AUTO_TEST_SUITE(Byte)
BOOST_AUTO_TEST_CASE(mem) {
    RAM[0x100] = 0x02;
    cpu.D[3] = 0x12345678;
    cpu.A[2] = 0x100;
    run_test(96);
    BOOST_TEST(RAM[0x100] == 0x78);
}

BOOST_AUTO_TEST_CASE(mem2mem) {
    RAM[0x100] = 0x02;
    RAM[0x200] = 0xff;
    run_test(100);
    BOOST_TEST(RAM[0x200] == 0x02);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)
BOOST_AUTO_TEST_CASE(mem) {
    TEST::SET_W(0x100, 0x0234);
    cpu.D[3] = 0x12345678;
    cpu.A[2] = 0x100;
    run_test(112);
    BOOST_TEST(TEST::GET_W(0x100) == 0x5678);
}

BOOST_AUTO_TEST_CASE(mem2mem) {
    TEST::SET_W(0x100, 0x1234);
    TEST::SET_W(0x200, 0x5678);
    run_test(116);
    BOOST_TEST(TEST::GET_W(0x200) == 0x1234);
}

BOOST_AUTO_TEST_CASE(CCR) {
    cpu.D[3] = 0x1F;
    run_test(128);
    BOOST_TEST(cpu.X);
    BOOST_TEST(cpu.Z);
    BOOST_TEST(cpu.V);
    BOOST_TEST(cpu.N);
    BOOST_TEST(cpu.C);
}

BOOST_AUTO_TEST_SUITE(SR)

BOOST_AUTO_TEST_CASE(NG) {
    cpu.S = false;
    run_test(132);
    BOOST_TEST(cpu.ex_n == 8);
}

BOOST_AUTO_TEST_CASE(OK) {
    cpu.S = true;
    cpu.D[3] = 0x1F;
    run_test(132);
    BOOST_TEST(!cpu.S);
    BOOST_TEST(cpu.X);
    BOOST_TEST(cpu.N);
    BOOST_TEST(cpu.V);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.Z);
}

BOOST_AUTO_TEST_CASE(Traced) {
    cpu.S = true;
    cpu.T = 1;
    cpu.D[3] = 1 << 14;
    run_test(132);
    BOOST_TEST(cpu.ex_n == 9);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)
BOOST_AUTO_TEST_CASE(mem) {
    TEST::SET_L(0x100, 0x02347654);
    cpu.D[3] = 0x12345678;
    cpu.A[2] = 0x100;
    run_test(136);
    BOOST_TEST(TEST::GET_L(0x100) == 0x12345678);
}

BOOST_AUTO_TEST_CASE(mem2mem) {
    TEST::SET_L(0x100, 0x12345678);
    TEST::SET_L(0x200, 0x56789ABC);
    run_test(140);
    BOOST_TEST(TEST::GET_L(0x200) == 0x12345678);
}
BOOST_AUTO_TEST_SUITE(USP)
BOOST_AUTO_TEST_CASE(NG) {
    cpu.S = false;
    run_test(152);
    BOOST_TEST(cpu.ex_n == 8);
}

BOOST_AUTO_TEST_CASE(OK) {
    cpu.S = true;
    cpu.A[3] = 0x300;
    run_test(152);
    BOOST_TEST(cpu.USP == 0x300);
}

BOOST_AUTO_TEST_CASE(Traced) {
    cpu.S = true;
    cpu.T = 1;
    run_test(152);
    BOOST_TEST(cpu.ex_n == 9);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/move.lua
-- MOVE.B %D2, %D3
MEM[0].W = OCT "010000" | 3 << 9 | 2
MEM[2].W = TEST.BREAK

-- MOVE.B (%A2), %D3
MEM[4].W = OCT "010020" | 3 << 9 | 2
MEM[6].W = TEST.BREAK

-- MOVE.B (%A2)+, %D3
MEM[8].W = OCT "010030" | 3 << 9 | 2
MEM[10].W = TEST.BREAK

-- MOVE.B (%A7)+, %D3
MEM[12].W = OCT "010030" | 3 << 9 | 7
MEM[14].W = TEST.BREAK

-- MOVE.B -(%A2), %D3
MEM[16].W = OCT "010040" | 3 << 9 | 2
MEM[18].W = TEST.BREAK

-- MOVE.B -(%A7), %D3
MEM[20].W = OCT "0010040" | 3 << 9 | 7
MEM[22].W = TEST.BREAK

-- MOVE.B #0x34, %D3
MEM[24].W = OCT "0010074" | 3 << 9
MEM[26].W = 0x34
MEM[28].W = TEST.BREAK

-- MOVE.W %D2, %D3
MEM[30].W = OCT "0030000" | 3 << 9 | 2
MEM[32].W = TEST.BREAK

-- MOVE.W %A2, %D3
MEM[34].W = OCT "0030010" | 3 << 9 | 2
MEM[36].W = TEST.BREAK

-- MOVE.W (%A2), %D3
MEM[38].W = OCT "0030020" | 3 << 9 | 2
MEM[40].W = TEST.BREAK

-- MOVE.W (%A2)+, %D3
MEM[42].W = OCT "0030030" | 3 << 9 | 2
MEM[44].W = TEST.BREAK

-- MOVE.W (%A7)+, %D3
MEM[46].W = OCT "0030030" | 3 << 9 | 7
MEM[48].W = TEST.BREAK

-- MOVE.W -(%A2), %D3
MEM[50].W = OCT "0030040" | 3 << 9 | 2
MEM[52].W = TEST.BREAK

-- MOVE.W -(%A7), %D3
MEM[54].W = OCT "0030040" | 3 << 9 | 7
MEM[56].W = TEST.BREAK

-- MOVE.W #0x3456, %D3
MEM[58].W = OCT "0030074" | 3 << 9
MEM[60].W = 0x3456
MEM[62].W = TEST.BREAK

-- MOVE.W %CCR, %D3
MEM[64].W = OCT "0041300" | 3
MEM[66].W = TEST.BREAK

-- MOVE.W %SR, %D3
MEM[68].W = OCT "0040300" | 3
MEM[70].W = TEST.BREAK

-- MOVE.L %D2, %D3
MEM[72].W = OCT "0020000" | 3 << 9 | 2
MEM[74].W = TEST.BREAK

-- MOVE.L %A2, %D3
MEM[76].W = OCT "0020010" | 3 << 9 | 2
MEM[78].W = TEST.BREAK

-- MOVE.L (%A2), %D3
MEM[80].W = OCT "0020020" | 3 << 9 | 2
MEM[82].W = TEST.BREAK

-- MOVE.L (%A2)+, %D3
MEM[84].W = OCT "0020030" | 3 << 9 | 2
MEM[86].W = TEST.BREAK

-- MOVE.L (%A7)+, %D3
MEM[88].W = OCT "0020030" | 3 << 9 | 7
MEM[90].W = TEST.BREAK

-- MOVE.L -(%A2), %D3
MEM[92].W = OCT "0020040" | 3 << 9 | 2
MEM[94].W = TEST.BREAK

-- MOVE.L -(%A7), %D3
MEM[96].W = OCT "0020040" | 3 << 9 | 7
MEM[98].W = TEST.BREAK

-- MOVE.L #0x34567890, %D3
MEM[100].W = OCT "0020074" | 3 << 9
MEM[102].L = 0x34567890
MEM[106].W = TEST.BREAK

-- MOVE.L %USP, %A3
MEM[108].W = OCT "0047150" | 3
MEM[110].W = TEST.BREAK

-- MOVE.B %D3, (%A2)
MEM[112].W = OCT "0010200" | 2 << 9 | 3
MEM[114].W = TEST.BREAK

-- MOVE.B (0x1000), (0x1200)
MEM[116].W = OCT "0011771"
MEM[118].L = 0x1000
MEM[122].L = 0x1200
MEM[126].W = TEST.BREAK

-- MOVE.W %D3, (%A2)
MEM[128].W = OCT "0030200" | 2 << 9 | 3
MEM[130].W = TEST.BREAK

-- MOVE.W (0x1000), (0x1200)
MEM[132].W = OCT "0031771"
MEM[134].L = 0x1000
MEM[138].L = 0x1200
MEM[142].W = TEST.BREAK

-- MOVE.W %D3, %CCR
MEM[144].W = OCT "0042300" | 3
MEM[146].W = TEST.BREAK

-- MOVE.W %D3, %SR
MEM[148].W = OCT "0043300" | 3
MEM[150].W = TEST.BREAK

-- MOVE.L %D3, (%A2)
MEM[152].W = OCT "0020200" | 2 << 9 | 3
MEM[154].W = TEST.BREAK

-- MOVE.L (0x1000), (0x1200)
MEM[156].W = OCT "0021771"
MEM[158].L = 0x1000
MEM[162].L = 0x1200
MEM[166].W = TEST.BREAK

-- MOVE.L %A3, %USP
MEM[168].W = OCT "0047140" | 3
MEM[170].W = TEST.BREAK
JIT_COMPILE(0, 172)

local run = {}

run.B = {
    n = farray(TF, function(b)
        CPU.D[3].B = cond(b, 0x80, 0)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[3].B = cond(b, 0, 1)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    dr = function()
        CPU.D[4].L = 0x12345678
        CPU.D[3].L = 0x9ABCDE02
        CPU.CC.C = true
        CPU.CC.V = true
        TEST.RUN(0)
        TEST.CHECK("CPU.D[4].L", 0x12345602)
        TEST.CHECK("CPU.CC.C", false)
        TEST.CHECK("CPU.CC.V", false)
    end,
    mem = function()
        MEM[0x1000].B = 0x34
        CPU.A[3] = 0x1000
        TEST.RUN(4)
        TEST.CHECK("CPU.D[4].B", 0x34)
    end,
    inc = function()
        MEM[0x1000].B = 0x56
        CPU.A[3] = 0x1000
        TEST.RUN(8)
        TEST.CHECK("CPU.D[4].B", 0x56)
        TEST.CHECK("CPU.A[3]", 0x1001)
    end,
    pop = function()
        MEM[0x1000].B = 0x78
        CPU.A[8] = 0x1000
        TEST.RUN(12)
        TEST.CHECK("CPU.D[4].B", 0x78)
        TEST.CHECK("CPU.A[8]", 0x1002)
    end,
    dec = function()
        MEM[0x1000].B = 0x9A
        CPU.A[3] = 0x1001
        TEST.RUN(16)
        TEST.CHECK("CPU.D[4].B", 0x9A)
        TEST.CHECK("CPU.A[3]", 0x1000)
    end,
    dec_sp = function()
        MEM[0x1000].B = 0x9B
        CPU.A[8] = 0x1002
        TEST.RUN(20)
        TEST.CHECK("CPU.D[4].B", 0x9B)
        TEST.CHECK("CPU.A[8]", 0x1000)
    end,
    imm = function()
        TEST.RUN(24)
        TEST.CHECK("CPU.D[4].B", 0x34)
    end,
    to_mem = function()
        CPU.D[4].B = 0xEF
        CPU.A[3] = 0x1000
        TEST.RUN(112)
        TEST.CHECK("MEM[0x1000].B", 0xEF)
    end,
    mem_mem = function()
        MEM[0x1000].B = 0xA9
        TEST.RUN(116)
        TEST.CHECK("MEM[0x1200].B", 0xA9)
    end
}

run.W = {
    n = farray(TF, function(b)
        CPU.D[3].W = cond(b, 0x8000, 0)
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[3].W = cond(b, 0, 1)
        TEST.RUN(30)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    dr = function()
        CPU.D[4].L = 0x12345678
        CPU.D[3].L = 0x9ABCDE02
        CPU.CC.C = true
        CPU.CC.V = true
        TEST.RUN(30)
        TEST.CHECK("CPU.D[4].L", 0x1234DE02)
        TEST.CHECK("CPU.CC.C", false)
        TEST.CHECK("CPU.CC.V", false)
    end,
    ar = function()
        CPU.D[4].L = 0x12345678
        CPU.A[3] = 0x9ABCDE02
        CPU.CC.C = true
        CPU.CC.V = true
        TEST.RUN(34)
        TEST.CHECK("CPU.D[4].L", 0x1234DE02)
    end,
    mem = function()
        MEM[0x1000].W = 0x1234
        CPU.A[3] = 0x1000
        TEST.RUN(38)
        TEST.CHECK("CPU.D[4].W", 0x1234)
    end,
    inc = function()
        MEM[0x1000].W = 0x5643
        CPU.A[3] = 0x1000
        TEST.RUN(42)
        TEST.CHECK("CPU.D[4].W", 0x5643)
        TEST.CHECK("CPU.A[3]", 0x1002)
    end,
    pop = function()
        MEM[0x1000].W = 0x7890
        CPU.A[8] = 0x1000
        TEST.RUN(46)
        TEST.CHECK("CPU.D[4].W", 0x7890)
        TEST.CHECK("CPU.A[8]", 0x1002)
    end,
    dec = function()
        MEM[0x1000].W = 0x9ABC
        CPU.A[3] = 0x1002
        TEST.RUN(50)
        TEST.CHECK("CPU.D[4].W", 0x9ABC)
        TEST.CHECK("CPU.A[3]", 0x1000)
    end,
    dec_sp = function()
        MEM[0x1000].W = 0x9BCD
        CPU.A[8] = 0x1002
        TEST.RUN(54)
        TEST.CHECK("CPU.D[4].W", 0x9BCD)
        TEST.CHECK("CPU.A[8]", 0x1000)
    end,
    imm = function()
        TEST.RUN(58)
        TEST.CHECK("CPU.D[4].W", 0x3456)
    end,
    cr = function()
        CPU.CC.value = 0x1F
        TEST.RUN(64)
        TEST.CHECK("CPU.D[4].W", 0x1F)
    end,
    sr = {
        ng = function() priv_test(68) end,
        ok = function()
            CPU.SR.value = 0x201F
            TEST.RUN(68)
            TEST.CHECK("CPU.D[4].W", 0x201F)
        end
    },
    to_mem = function()
        CPU.D[4].W = 0x4532
        CPU.A[3] = 0x1000
        TEST.RUN(128)
        TEST.CHECK("MEM[0x1000].W", 0x4532)
    end,
    mem_mem = function()
        MEM[0x1000].W = 0x4258
        TEST.RUN(132)
        TEST.CHECK("MEM[0x1200].W", 0x4258)
    end,
    to_cr = function()
        CPU.D[4].W = 0x1F
        TEST.RUN(144)
        TEST.CHECK("CPU.CC.value", 0x1F)
    end,
    to_sr = {
        ng = function() priv_test(148) end,
        ok = function()
            CPU.SR.S = true
            CPU.D[4].W = 0x201F
            TEST.RUN(148)
            TEST.CHECK("CPU.SR.value", 0x201F)
        end
    },
}

run.L = {
    n = farray(TF, function(b)
        CPU.D[3].L = cond(b, 0x80000000, 0)
        TEST.RUN(72)
        TEST.CHECK("CPU.CC.N", b)
    end),
    z = farray(TF, function(b)
        CPU.D[3].L = cond(b, 0, 1)
        TEST.RUN(72)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    dr = function()
        CPU.D[4].L = 0x12345678
        CPU.D[3].L = 0x9ABCDE02
        CPU.CC.C = true
        CPU.CC.V = true
        TEST.RUN(72)
        TEST.CHECK("CPU.D[4].L", 0x9ABCDE02)
        TEST.CHECK("CPU.CC.C", false)
        TEST.CHECK("CPU.CC.V", false)
    end,
    ar = function()
        CPU.D[4].L = 0x12345678
        CPU.A[3] = 0x9ABCDE02
        CPU.CC.C = true
        CPU.CC.V = true
        TEST.RUN(76)
        TEST.CHECK("CPU.D[4].L", 0x9ABCDE02)
    end,
    mem = function()
        MEM[0x1000].L = 0x12345678
        CPU.A[3] = 0x1000
        TEST.RUN(80)
        TEST.CHECK("CPU.D[4].L", 0x12345678)
    end,
    inc = function()
        MEM[0x1000].L = 0x5643210F
        CPU.A[3] = 0x1000
        TEST.RUN(84)
        TEST.CHECK("CPU.D[4].L", 0x5643210F)
        TEST.CHECK("CPU.A[3]", 0x1004)
    end,
    pop = function()
        MEM[0x1000].L = 0x78901234
        CPU.A[8] = 0x1000
        TEST.RUN(88)
        TEST.CHECK("CPU.D[4].L", 0x78901234)
        TEST.CHECK("CPU.A[8]", 0x1004)
    end,
    dec = function()
        MEM[0x1000].L = 0x9ABCDEF0
        CPU.A[3] = 0x1004
        TEST.RUN(92)
        TEST.CHECK("CPU.D[4].L", 0x9ABCDEF0)
        TEST.CHECK("CPU.A[3]", 0x1000)
    end,
    dec_sp = function()
        MEM[0x1000].L = 0x9BCDEF01
        CPU.A[8] = 0x1004
        TEST.RUN(96)
        TEST.CHECK("CPU.D[4].L", 0x9BCDEF01)
        TEST.CHECK("CPU.A[8]", 0x1000)
    end,
    imm = function()
        TEST.RUN(100)
        TEST.CHECK("CPU.D[4].L", 0x34567890)
    end,
    usp = {
        ng = function() priv_test(108) end,
        ok = function()
            CPU.SR.S = true
            CPU.SR.T = 0
            CPU.USP = 0x5F00
            TEST.RUN(108)
            TEST.CHECK("CPU.A[4]", 0x5F00)
        end
    },
    to_mem = function()
        CPU.D[4].L = 0x45329754
        CPU.A[3] = 0x1000
        TEST.RUN(152)
        TEST.CHECK("MEM[0x1000].L", 0x45329754)
    end,
    mem_mem = function()
        MEM[0x1000].L = 0x42581234
        TEST.RUN(156)
        TEST.CHECK("MEM[0x1200].L", 0x42581234)
    end,
    to_usp = {
        ng = function() priv_test(168) end,
        ok = function()
            CPU.SR.S = true
            CPU.A[4] = 0x3000
            TEST.RUN(168)
            TEST.CHECK("CPU.USP", 0x3000)
        end
    },
}
return run
